sample_size/sample_size_calculator.py

```diff
@@ -0,0 +1,81 @@
+from typing import List
+from typing import Union
+
+import numpy as np
+
+from sample_size.metrics import BooleanMetric
+from sample_size.metrics import NumericMetric
+from sample_size.metrics import RatioMetric
+
+DEFAULT_ALPHA = 0.05
+DEFAULT_POWER = 0.8
+
+
+class SampleSizeCalculator:
+    """This class is to calculate sample size based on variable type
+
+    Attributes:
+    alpha: statistical significance
+    power: statistical power
+
+    """
+
+    def __init__(self, alpha: float = DEFAULT_ALPHA, power: float = DEFAULT_POWER):
+        self.alpha = alpha
+        self.power = power
+        # Consider having a self.metrics to hold all metric types
+        self.boolean_metrics: List[BooleanMetric] = []
+        self.numeric_metrics: List[NumericMetric] = []
+        self.ratio_metrics: List[RatioMetric] = []
+
+    def register_bool_metric(self, probability: float, mde: float) -> None:
+        metric = BooleanMetric(probability, mde)
+        self.boolean_metrics.append(metric)
+
+    def register_numeric_metric(self, variance: float, mde: float) -> None:
+        metric = NumericMetric(variance, mde)
+        self.numeric_metrics.append(metric)
+
+    def register_ratio_metric(
+        self,
+        numerator_mean: float,
+        numerator_variance: float,
+        denominator_mean: float,
+        denominator_variance: float,
+        covariance: float,
+        mde: float,
+    ) -> None:
+        metric = RatioMetric(
+            numerator_mean, numerator_variance, denominator_mean, denominator_variance, covariance, mde
+        )
+        self.ratio_metrics.append(metric)
+
+    def _get_single_sample_size(self, metric: Union[BooleanMetric, NumericMetric, RatioMetric]) -> float:
+        effect_size = metric.mde / float(np.sqrt(metric.variance))
+        power_analysis = metric.default_power_analysis_type()
+        sample_size = int(
+            power_analysis.solve_power(
+                effect_size=effect_size,
+                alpha=self.alpha,
+                power=self.power,
+                ratio=1,
+                alternative="two-sided",
+            )
+        )
+        return sample_size
+
+    def get_overall_sample_size(self) -> float:
```

Comment:
should this just be `get_sample_size`? or do we plan to expose multiple methods, one for overall, one for individual?

Reply:
done